### PR TITLE
Improve crate documentation following Rust guidelines

### PR DESCRIPTION
**Pull request by CPerezz**
This mainly focuses on:
- Extend the documentation of the crate exported structs,
traits & it's associated methods.
- Improve the quality of the previous existing docs.
- Add doc code examples as docs.
- Adapt the docs to the Rust documentation guidelines

Resolves: #505

This PR is made on the top of #507 and therefore has to be merged after it is merged.

### CODE REVIEW

File: src/circuit.rs [hunk 1]
```diff
@@ -54,11 +57,15 @@ pub struct VerifierData {
 impl VerifierData {
     /// Creates a new [`VerifierData`] from a [`VerifierKey`] and the public
     /// input positions of the circuit that it represents.
+    ///
+    /// [`VerifierKey`]: struct.VerifierKey.html
```
Comment (ZER0) [inline src/circuit.rs:61]: Not necessary.

File: src/circuit.rs [hunk 2]
```diff
@@ -54,11 +57,15 @@ pub struct VerifierData {
...
     pub const fn new(key: VerifierKey, pi_pos: Vec<usize>) -> Self {
         Self { key, pi_pos }
     }
 
     /// Returns a reference to the contained [`VerifierKey`].
+    ///
+    /// [`VerifierKey`]: struct.VerifierKey.html
```
Comment (ZER0) [inline src/circuit.rs:68]: Not necessary.

**Review by ZER0 — CHANGES_REQUESTED** [review]
I stopped to review and test since we have either two scenarios:
1. The additional markdown for links is unnecessary and sometimes actually harmful since it points to the wrong page (therefore those addition you made can be reverted)
2. Your local distribution of `cargo doc` differs from mine when it generates such links. 

So before proceed to comment everything, let's double check in which scenario are we (for example, check the links of the types I've commented as "harmful" and see if they lead to a wrong page, and then try to remove it and check if they're working).

**Review by vlopes11 — APPROVED** [review]
LGTM

**Review by ZER0 — APPROVED** [review]

**Pull request merged**